zerver/lib/export.py

```diff
@@ -0,0 +1,653 @@
+            return 0
+        for (status, job) in run_parallel(run_job, [str(x) for x in range(0, threads)], threads=threads):
+            print("Shard %s finished, status %s" % (job, status))
+            pass
+    logging.info("Finished exporting %s" % (realm.domain))
+
+def do_export_user(user_profile, output_dir):
+    response = {} # type: Dict[str, Any]
+
+    export_single_user(user_profile, response)
+    export_file = os.path.join(output_dir, "user.json")
+    with open(export_file, "w") as f:
+        f.write(ujson.dumps(response, indent=4))
+    logging.info("Exporting messages")
+    export_messages_single_user(user_profile, output_dir=output_dir)
+
+def export_single_user(user_profile, response):
+    response['zerver_userprofile'] = [model_to_dict(x, exclude=["password", "api_key"])
+                                      for x in [user_profile]]
+    floatify_datetime_fields(response, 'zerver_userprofile', 'date_joined')
+    floatify_datetime_fields(response, 'zerver_userprofile', 'last_login')
+    floatify_datetime_fields(response, 'zerver_userprofile', 'last_reminder')
+
+    response["zerver_subscription"] = [model_to_dict(x) for x in
+                                       Subscription.objects.filter(user_profile=user_profile)]
+    recipient_ids = set(s["recipient"] for s in response["zerver_subscription"])
+    response["zerver_recipient"] = [model_to_dict(x) for x in
+                                    Recipient.objects.filter(id__in=recipient_ids)]
+    stream_ids = set(x["type_id"] for x in response["zerver_recipient"] if x["type"] == Recipient.STREAM)
+
+    stream_query = Stream.objects.filter(id__in=stream_ids)
+    response['zerver_stream'] = [model_to_dict(x, exclude=["email_token"]) for x in stream_query]
+    floatify_datetime_fields(response, 'zerver_stream', 'date_created')
+
+def export_messages_single_user(user_profile, chunk_size=1000, output_dir=None):
+    user_message_query = UserMessage.objects.filter(user_profile=user_profile)
+    min_id = -1
+    dump_file_id = 1
+    while True:
+        actual_query = user_message_query.select_related("message", "message__sending_client").filter(id__gt=min_id)[0:chunk_size]
+        user_message_chunk = [um for um in actual_query]
+        user_message_ids = set(um.id for um in user_message_chunk)
+
+        if len(user_message_chunk) == 0:
+            break
+
+        message_chunk = []
+        for user_message in user_message_chunk:
+            item = model_to_dict(user_message.message)
+            item['flags'] = user_message.flags_list()
+            item['flags_mask'] = user_message.flags.mask
+            # Add a few nice, human-readable details
+            item['sending_client_name'] = user_message.message.sending_client.name
+            item['display_recipient'] = get_display_recipient(user_message.message.recipient)
+            message_chunk.append(item)
+
+        message_filename = os.path.join(output_dir, "messages-%04d.json" % (dump_file_id,))
```

Comment:
Use 5 or 6 digit formatting instead of 4, just in case we have more than 10k chunks.